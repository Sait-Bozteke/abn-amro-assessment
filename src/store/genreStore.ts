import { defineStore } from "pinia";
import { Show } from "../types/Show";

export const useGenreStore = defineStore('genreStore', {
    state: ()=>({
        genres : ['Drama', 'Comedy', 'Action']
    })
})