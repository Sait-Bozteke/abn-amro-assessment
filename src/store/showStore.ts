import { defineStore } from 'pinia';
import { ref } from 'vue';
import {
  getAllShows,
  getShowById,
  searchShowByName,
} from '../services/showService';
import { Show } from '../types/Show';

export const useShowStore = defineStore('showStore', () => {
  const shows = ref<Show[]>([]);
  const selectedShow = ref<Show | null>(null);
  const loading = ref(false);
  const error = ref<string | null>(null);

  const fetchAllShows = async () => {
    loading.value = true;
    error.value = null;
    try {
      shows.value = await getAllShows();
    } catch {
      error.value = 'Failed to fetch shows';
    } finally {
      loading.value = false;
    }
  };

  const fetchShowById = async (id: number) => {
    loading.value = true;
    error.value = null;
    try {
      selectedShow.value = await getShowById(id);
    } catch (err) {
      console.error(err);
      error.value = 'Failed to fetch show details';
    } finally {
      loading.value = false;
    }
  };

  const searchShows = async (query: string) => {
    loading.value = true;
    error.value = null;
    try {
      shows.value = await searchShowByName(query);
    } catch {
      error.value = 'Failed to search shows';
    } finally {
      loading.value = false;
    }
  };

  return {
    shows,
    selectedShow,
    loading,
    error,
    fetchAllShows,
    fetchShowById,
    searchShows,
  };
});
