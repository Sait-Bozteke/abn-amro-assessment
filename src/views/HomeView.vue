<template>
  <div class="container my-4">
    <h1 class="text-center mb-4">Select a Genre</h1>
    <div class="row">
      <CategoryCard
        v-for="genre in genres"
        :key="genre"
        :genre="genre"
        class="col-md-6 col-lg-4 mb-4"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import CategoryCard from "../components/CategoryCard.vue";
import { useGenreStore } from "../store/genreStore";

export default defineComponent({
  components: {
    CategoryCard,
  },
  setup() {
    const genreStore = useGenreStore();
    return {
      genres: genreStore.genres,
    };
  },
});
</script>

<style scoped>
.card {
  text-decoration: none;
  color: inherit;
  transition: transform 0.3s;
}
.card:hover {
  transform: scale(1.05);
}
</style>
