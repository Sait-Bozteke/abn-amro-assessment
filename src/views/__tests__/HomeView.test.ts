import { shallowMount } from '@vue/test-utils';
import HomeView from '@/views/HomeView.vue';
import { useGenreStore } from '../../store/genreStore';

// Mock the genre store
jest.mock('../../store/genreStore', () => ({
  useGenreStore: jest.fn(),
}));

describe('HomeView.vue', () => {
  it('renders correctly and matches snapshot', () => {
    (useGenreStore as unknown as jest.Mock).mockReturnValue({
      genres: ['Comedy', 'Drama', 'Action'],
    });

    const wrapper = shallowMount(HomeView);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders the correct number of CategoryCard components', () => {
    (useGenreStore as unknown as jest.Mock).mockReturnValue({
      genres: ['Comedy', 'Drama', 'Action'],
    });

    const wrapper = shallowMount(HomeView);
    expect(wrapper.findAllComponents({ name: 'CategoryCard' }).length).toBe(3);
  });
});
