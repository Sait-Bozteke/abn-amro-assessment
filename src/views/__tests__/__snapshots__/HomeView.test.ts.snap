// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeView.vue renders correctly and matches snapshot 1`] = `
"<div class="container my-4">
  <h1 class="text-center mb-4">Select a Genre</h1>
  <div class="row">
    <category-card-stub genre="Comedy" class="col-md-6 col-lg-12"></category-card-stub>
    <category-card-stub genre="Drama" class="col-md-6 col-lg-12"></category-card-stub>
    <category-card-stub genre="Action" class="col-md-6 col-lg-12"></category-card-stub>
  </div>
</div>"
`;
